patch/llpcPatchResourceCollect.cpp

```diff
@@ -2372,6 +2402,385 @@ void PatchResourceCollect::MapGsBuiltInOutput(
+            {
+                // i8/i16 -> i32
+                if (pCompTy->getScalarSizeInBits() < 32)
+                {
+                    pComp = builder.CreateZExt(pComp, m_pContext->Int32Ty());
+                }
+                // i32 -> float
+                pComp = builder.CreateBitCast(pComp, m_pContext->FloatTy());
+            }
+            else if (pCompTy->isHalfTy())
+            {
+                // f16 -> float
+                pComp = builder.CreateFPExt(pComp, m_pContext->FloatTy());
+            }
+
+            if (compCount > 1)
+            {
+                Value* pCompIdx = ConstantInt::get(m_pContext->Int32Ty(), compIdx);
+                pOutValue = builder.CreateInsertElement(pOutValue, pComp, pCompIdx);
+            }
+            else
+            {
+                pOutValue = pComp;
+            }
+        }
+
+        args[0] = ConstantInt::get(m_pContext->Int32Ty(), consectiveLocation);
+        args[1] = ConstantInt::get(m_pContext->Int32Ty(), 0);
+        args[2] = pOutValue;
+
+        std::string callName(LlpcName::OutputExportGeneric);
+        AddTypeMangling(m_pContext->VoidTy(), args, callName);
+
+        EmitCall(callName, m_pContext->VoidTy(), args, NoAttrib, builder);
+
+        outputLocMap[consectiveLocation] = InvalidValue;
+        ++consectiveLocation;
+    }
+}
+
+// =====================================================================================================================
+// Fill the locationSpan container by constructing a LocationSpan from each input import call
+bool InOutLocationMapManager::AddSpan(
+    CallInst*   pCall)  // [in] Call to process
+{
+    auto pCallee = pCall->getCalledFunction();
+    auto mangledName = pCallee->getName();
+    bool isInput = false;
+    if (mangledName.startswith(LlpcName::InputImportGeneric))
+    {
+        LocationSpan span = {};
+
+        span.firstLocation.locationInfo.location = cast<ConstantInt>(pCall->getOperand(0))->getZExtValue();
+        span.firstLocation.locationInfo.component = cast<ConstantInt>(pCall->getOperand(1))->getZExtValue();
+        span.firstLocation.locationInfo.half = false;
+
+        const uint32_t bitWidth = pCallee->getReturnType()->getScalarSizeInBits();
+        span.compatibilityInfo.halfComponentCount = bitWidth < 64 ? 2 : 4;
+
+        span.compatibilityInfo.isFlat = (cast<ConstantInt>(pCall->getOperand(2))->getZExtValue() == InterpModeFlat);
+        span.compatibilityInfo.is16Bit = false;
+        span.compatibilityInfo.isCustom =
+            (cast<ConstantInt>(pCall->getOperand(2))->getZExtValue() == InterpModeCustom);
+
+        LLPC_ASSERT(std::find(m_locationSpans.begin(), m_locationSpans.end(), span) == m_locationSpans.end());
+        m_locationSpans.push_back(span);
+
+        isInput = true;
+    }
+    if (mangledName.startswith(LlpcName::InputImportInterpolant))
+    {
+        auto pLocOffset = pCall->getOperand(1);
+        if (isa<ConstantInt>(pLocOffset))
+        {
```

Comment:
What happens if pLocOffset is not constant?

Reply:
In an early change of supporting dynamic index in array with interpolant, we create `@llpc.input.import.interpolant` for each element in the array in the lower global pass. (see InterpolateInputElement()). So I think pLocOffset is always contant after that change. I will add an assertion here first and clean the check of pLocOffset in the patch pass later. What do you think?

Reply:
Yes, that sounds good to me!